sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,283 @@
+    obj = _obj;
+  }
+#endif
+
+public:
+  static_assert(std::is_trivially_destructible<T>::value,
+                "Type T must be trivially destructible.");
+  static_assert(is_property_list<property_list_t>::value,
+                "Property list is invalid.");
+  // static_assert(detail::SortedAllUnique<std::tuple<Props...>>::value,
+  // "Duplicate properties in property list.");
+
+  annotated_arg() noexcept = default;
+  annotated_arg(const annotated_arg &) = default;
+  annotated_arg &operator=(annotated_arg &) = default;
+
+  annotated_arg(T *_ptr,
+                const property_list_t &PropList = properties{}) noexcept
+      : obj(sycl::detail::cast_AS<global_pointer_t>(_ptr)) {}
+
+  template <typename... PropertyValueTs>
+  annotated_arg(T *_ptr, PropertyValueTs... props) noexcept
+      : obj(sycl::detail::cast_AS<global_pointer_t>(_ptr)) {
+    static_assert(
+        std::is_same<property_list_t,
+                     detail::merged_properties_t<
+                         property_list_t,
+                         detail::properties_t<PropertyValueTs...>>>::value,
+        "The property list must contain all properties of the input of the "
+        "constructor");
+  }
+
+  // Constructs an annotated_arg object from another annotated_arg object.
+  // The property set PropertyListT contains all properties of the input
+  // annotated_arg object. If there are duplicate properties present in the
+  // property list of the input annotated_arg object, the values of the
+  // duplicate properties must be the same.
+  template <typename T2, typename PropertyList2>
+  explicit annotated_arg(const annotated_arg<T2, PropertyList2> &other) noexcept
+      : obj(other.obj) {
+    static_assert(std::is_convertible<T2, T *>::value,
+                  "The underlying data type of the input annotated_arg is not "
+                  "compatible");
+
+    static_assert(
+        std::is_same<
+            property_list_t,
+            detail::merged_properties_t<property_list_t, PropertyList2>>::value,
+        "The constructed annotated_arg type must contain all the properties of "
+        "the input annotated_arg");
+  }
+
+  template <typename T2, typename PropertyListU, typename PropertyListV>
+  explicit annotated_arg(const annotated_arg<T2, PropertyListU> &other,
+                         const PropertyListV &proplist) noexcept
+      : obj(other.obj) {
+    static_assert(std::is_convertible<T2, T *>::value,
+                  "The underlying data type of the input annotated_arg is not "
+                  "compatible");
+
+    static_assert(
+        std::is_same<property_list_t, detail::merged_properties_t<
+                                          PropertyListU, PropertyListV>>::value,
+        "The property list of constructed annotated_arg type must be the union "
+        "of the input property lists");
+  }
+
+  operator T *() const noexcept { return obj; }
```

Comment:
Note: corresponding method in the spec is not marked as `const`

Reply:
Removed